library/core/src/iter/traits/iterator.rs

```diff
@@ -48,13 +48,24 @@ pub trait Iterator {
     type Item;
 
     /// Advances the iterator and returns the next value.
+    /// Returns [`None`] when iteration is finished.
     ///
-    /// Returns [`None`] when iteration is finished. Individual iterator
-    /// implementations may choose to resume iteration, and so calling `next()`
-    /// again may or may not eventually start returning [`Some(Item)`] again at some
-    /// point.
+    /// # Behavior after [`None`] is returned
+    ///
+    /// Calling `next()` on an iterator after it has returned `None`
+    /// is in general a perfectly valid use of the interface.
```

Comment:
I agree that this is not worded in the most understandable way. I think the point of this sentence is to say calling `next` after `None` is not defined by Iterator (on behalf of all impls) as a misuse of the interface. This is unlike Future where polling a previously ready future is defined as misuse by default, absent some other documentation from a specific impl such as futures::future::Fuse. So it is not something that every Future impl is responsible for considering; "whatever happens, happens" is fine at least within the bounds of memory safety. In contrast, we are saying all Iterator impls _are_ responsible for considering how they behave in this case, and there are valid expectations in excess of just being memory-safe.

If this is the point being made, "in general perfectly valid" isn't a good way to frame it because it isn't so valid in generic code that operates on any arbitrary iterator, as you've noted here, and as in your example from https://github.com/rust-lang/rust/pull/123692#issuecomment-2046773681.